### ISSUE TITLE
Allow user to process data

### ISSUE THREAD

**Issue by woodthom2** [harmonydata/harmony#40]
## Description

At the end of the Harmony user journey when the user exports results as xlsx (  https://youtu.be/CqAsrY74zNM ), can we generate either some Python code or a Colab or Jupyter notebook allowing them to analyse their datasets?

## Rationale

This would be a nice feature to streamline the whole data harmonisation process. 

We could offer the option to do it in the Web UI but by helping the user complete it on their machine we bypass some confidentiality issues (if user is not allowed to upload raw data to internet)

An open question: how can the user do a statistical analysis and incorporate the Harmony scores? Do we make a new variable which is e.g. 0.65 × Instr1Ques1 + 0.33 × Instr2Ques5 etc... and then do statistical tests on it???

**Comment by Bpriya42** [comment]
Hey @woodthom2, Is this feature still a priority? I would like to contribute to this issue. Please let me know. Thanks :)

**Comment by woodthom2** [comment]
Yes please go ahead! Thanks. Do you want to have a conversation to discuss
how to do it? You can find me on the Harmony Discord
https://discord.gg/harmonydata

**Comment by Bpriya42** [comment]
**Plan of Action for this feature**

1. Z‑score each item to put all question responses on a common scale.

2. Compute item strength by summing its Harmony match_scores to every other item in its cluster.

3. Normalize weights by dividing each item’s strength by the cluster’s total strength (so all weights sum to 1).

4. Form the composite by multiplying each respondent’s item z‑score by its normalized weight and summing across the cluster.

Currently, we are assuming that the results will be stored in a directory that can be set up by the user locally.  This solution will require both R and Python solutions. 

Hey @woodthom2 , please let me know if this covers everything? 
I can work on a solution for Python

**Comment by woodthom2** [comment]
Hi @Bpriya42 ! Thanks so much, I have some thoughts about this after discussing with everyone else on the project. I will try to get some more feedback from them.

I'm wondering if things like Z-scores are steps that we should leave to the user. Perhaps the user has their own stats technique. I would like to be careful about extending the scope too much as stats packages already exist. But I am not sure about this. However maybe Z-score is the best approach. I would like to get feedback /ideas from colleagues @bmoltrecht @eoinmcelroy and @ronnyTodgers 

Some of the feedback I have got so far is "this will not be the whole process, data harmonisation can also include things like harmonising the structures of different datasets (eg one might be wide -and other long) and it will also depend on the type of variables."

Also it's worth checking what happens in the case of negative meaning (where items are opposites, and Harmony would give a negative value)

For example, my colleague Eoin McElroy (@eoinmcelroy) has written some write-ups of the harmonisation process, I am wondering if your suggestion tallies with these?

* https://www.closer.ac.uk/wp-content/uploads/Data-harmonisation-A3-McElroy-Eoin.pdf
* https://www.closer.ac.uk/wp-content/uploads/210715-Harmonisation-measurement-properties-mental-health-measures-british-cohorts.pdf

I will try to get some more information to you in the next few days or get a stats expert to give me some input on your ideas

**Comment by ronnyTodgers** [comment]
I would just say that the approach outlined seems to be a little over interpretive in terms of using the semantic clusters weights. Generally if we define a cluster of items in a scale as forming a subscale we would not then weight those items differently to compute a subscale metric. 
It seems the approach as outlined would not only use the clusters to make subscales, but go further and weight the items within those subscales. This undoubtedly makes the derived subscale variable better reflect the cluster - I would not dispute that. But it feels over interpretive. As a minimum I would suggest offering a simple subscale metric calculation alongside this!

But yes I would also echo the interest in whether the current cluster can be extended to deal with the "I'm very happy / I'm very unhappy" being clustered as you would expect to happen based on a data approach with reverse coding applied.

**Comment by Bpriya42** [comment]
Hey @woodthom2 and @ronnyTodgers, Thank you for your suggestions. I have come up with a solution, please let me know if this works. The aim is to have a working solution that can be iterated on later, so here is the initial approach. 

### Problem

- Researchers need a ready-to-run notebook that works locally to respect confidentiality constraints.
- The notebook should focus on transparent, reproducible harmonisation steps rather than statistical weighting or modelling. 

### Solution

- The solution isinspired by the steps taken in this[ paper ](https://acamh.onlinelibrary.wiley.com/doi/full/10.1111/jcpp.14080)to complete the harmonisation of data from the UK Millennium Cohort Study (MCS) and the Brazilian High Risk Cohort Study (BHRCS). 
- Instead of adding statistical complexity (e.g. z-scores), we can remap the responses to a scale that the user wants.

1. Variable Mapping: Here the users can review the matched item pairs (that Harmony already completes) and provide a mapping table that specifies:
- Variable names per study
- Harmonised concept/label
- Response recoding/binarisation rule

2. Interactive code cells for flexibility:
- Pre-filled notebook cells show how to apply the mappings.
- Instructions explain how to adjust mappings.

3. Apply Mapping rules: This will allow us to harmonise responses for each "cohort" or matching pairs. After this step, each item will be represented as either a binary score (as suggested in the paper) or another scale, depending on the user’s mapping choices.

4. Combine data and Export: Harmonised dataframes from all the cohorts can then be stacked into one dataframe with cohort labels specified. We can then provide the option to export this dataframe as csv, xlsx, or other formats. 

For this initial version, the focus is on harmonising categorical values. Looking ahead, we should ensure that the same mapping logic can also accommodate continuous variables. 

Please let me know if this solution looks good.
